tchannel-core/src/main/java/com/uber/tchannel/messages/ThriftRequest.java

```diff
@@ -58,14 +58,43 @@ public Builder(String service, ByteBuf arg1) {
             this.argScheme = ArgScheme.THRIFT;
         }
 
+        /**
+         * Validates payload and populates {@link #arg1}, {@link #arg2}, {@link #arg3}.
+         *
+         * Args <b>>need</b> to be cleared if validation fails.
+         *
+         * Use {@link #release()} to clear args above.
+         *
+         * @throws Exception
+         *     if validation fails.
+         */
         @Override
         public Builder<T> validate() {
             super.validate();
             return this;
         }
 
+        /**
+         * Validates payload, populates {@link #arg1}, {@link #arg2}, {@link #arg3} and builds {@link ThriftRequest}.
+         *
+         * Args above will be auto-released if validation fails.
+         *
+         * Note: Don't call it again if fails.
+         * <br/>unless header/body are re-populated this method will fail if called after the initial failure.
+         */
         public ThriftRequest<T> build() {
-            return new ThriftRequest<>(this.validate());
+            ThriftRequest<T> result;
```

Comment:
could we create a copy of the builder and call build on that? this will allow reusing the builder even if validation fails.

Reply:
I made Request.validateHeader and Request.validateBody idempotent: they can be called number of times w/o producing side effect if arg2/arg3 already constructed.